src/engine/sparqlExpressions/SparqlExpressionGenerators.h

```diff
@@ -200,6 +202,166 @@ CPP_template(typename Input, typename Transformation = ql::identity)(
   }
 }
 
+// A wrapper range that takes an underlying generator/range of values for all
+// rows in a block and a sorted range of row indices and yields only the
+// values at those indices. The underlying generator is advanced without
+// dereferencing for the skipped indices.
+template <typename GeneratorRange, typename IndicesRange>
+class SparseGeneratorRange {
+ private:
+  GeneratorRange generator_;
+  IndicesRange indices_;
+  size_t totalSize_;
+
+ public:
+  using value_type = ql::ranges::range_value_t<GeneratorRange>;
+
+  SparseGeneratorRange(GeneratorRange generator, IndicesRange indices,
+                       size_t totalSize)
+      : generator_{std::move(generator)},
+        indices_{std::move(indices)},
+        totalSize_{totalSize} {}
+
+  class iterator {
+   private:
+    using GenRange = GeneratorRange;
+    using IdxRange = IndicesRange;
+
+    GenRange* generator_ = nullptr;
+    IdxRange* indices_ = nullptr;
+    ql::ranges::iterator_t<GenRange> genIt_{};
+    ql::ranges::sentinel_t<GenRange> genEnd_{};
+    ql::ranges::iterator_t<IdxRange> idxIt_{};
+    ql::ranges::sentinel_t<IdxRange> idxEnd_{};
+    size_t currentIndex_ = 0;
+    size_t totalSize_ = 0;
+
+    void skipToCurrentTarget() {
```

Comment:
make the previous target and current target parameters here, then this gets much simpler.